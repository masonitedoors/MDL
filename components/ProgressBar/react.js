import React, { useReducer, useEffect } from 'react'
import classNames from 'classnames/bind'
import PropTypes from 'prop-types'
import styles from './style.module.scss'

const cx = classNames.bind(styles)

const reducer = (state, { type, payload }) => {
  switch (type) {
    case 'setFillLength':
      return {
        ...state,
        fillLength: payload.fillLength,
        fillTransition: Object.prototype.hasOwnProperty.call(payload, 'fillTransition')
          ? payload.fillTransition
          : true,
      }
    case 'setFillIndex':
      return { ...state, fillIndex: payload }
    case 'fillToIndex':
      return (() => {
        const fillIndex = state.fillToIndex > state.prevFillToIndex ? state.fillIndex + 1 : state.fillIndex - 1
        return {
          ...state,
          fillIndex: fillIndex > 0 ? fillIndex : 0,
        }
      })()
    case 'setFillToIndex':
      return {
        ...state,
        fillToIndex: payload > -1 ? payload : 0,
        prevFillToIndex: state.fillToIndex,
      }
    case 'updatePointData':
      return {
        ...state,
        pointOffsetStarts: payload.pointOffsetStarts || state.pointOffsetStarts,
        pointOffsetSizes: payload.pointOffsetSizes || state.pointOffsetSizes,
      }
    default:
      throw new Error()
  }
}

const PROGRESS_POINT_SIZE = 16

const initialState = {
  fillLength: '8px',
  fillIndex: 0,
  fillToIndex: 0,
  fillTransition: true,
  prevFillToIndex: null,
  pointOffsetSizes: [],
  pointOffsetStarts: [],
}

const ProgressBar = React.forwardRef(
  ({
    progressPoints, currentPointId, horizontal, recalculate,
  }, ref) => {
    const progressPointRefs = []

    const [state, dispatch] = useReducer(reducer, initialState)
    const {
      fillLength,
      fillIndex,
      fillToIndex,
      fillTransition,
      pointOffsetSizes,
      pointOffsetStarts,
    } = state

    const currentPointIdIndex = progressPoints && progressPoints.map(({ id }) => id).indexOf(currentPointId)

    const points = progressPoints && progressPoints.map((progressPoint, index) => {
      progressPointRefs.push(React.createRef())

      return (
        <div
          className={cx('progress-bar__point-label')}
          key={progressPoint.id}
          ref={progressPointRefs[index]}
        >
          <div
            className={cx('progress-bar__point', {
              'progress-bar__point--active': currentPointIdIndex >= index && fillIndex >= index,
            })}
          >
            {progressPoint.render && progressPoint.render()}
          </div>
        </div>
      )
    })

    const progressBarMeterStyle = {
      [horizontal ? 'right' : 'bottom']: `${pointOffsetSizes[pointOffsetSizes.length - 1]
        - PROGRESS_POINT_SIZE / 2}px`,
    }

    const barFillStyle = {
      [horizontal ? 'width' : 'height']: fillLength,
    }

    /*
     * Inform progress bar of the fill target for staggered fill animation.
     */
    useEffect(() => {
      dispatch({ type: 'setFillToIndex', payload: currentPointIdIndex })
    }, [currentPointIdIndex])

    /*
     * In case the initial currentPointId is not the first, let's default this.
     */
    useEffect(() => {
      dispatch({ type: 'setFillIndex', payload: currentPointIdIndex })
    }, [])

    /*
     * Set progress bar fill length.
     */
    useEffect(() => {
      if (
        progressPointRefs.length > 0
        && progressPointRefs.length === progressPoints.length
        && fillIndex !== fillToIndex
      ) {
        const direction = fillToIndex > fillIndex ? 1 : -1

        const pointOffsetStarts = progressPointRefs.map(point =>
          (horizontal ? point.current.offsetLeft : point.current.offsetTop))
        const pointOffsetSizes = progressPointRefs.map(point =>
          (horizontal ? point.current.offsetWidth : point.current.offsetHeight))

        dispatch({
          type: 'updatePointData',
          payload: { pointOffsetStarts, pointOffsetSizes },
        })
        dispatch({
          type: 'setFillLength',
          payload: {
            fillLength: `${
              pointOffsetStarts[fillIndex + direction > 0 ? fillIndex + direction : 0]
            }px`,
          },
        })
      }
    }, [pointOffsetStarts.join(''), fillToIndex, fillIndex])

    /*
     * Force recalculations of progress bar (such as when a point's content's size changes.
     */
    useEffect(() => {
      const pointOffsetStarts = progressPointRefs.map(point =>
        (horizontal ? point.current.offsetLeft : point.current.offsetTop))
      const pointOffsetSizes = progressPointRefs.map(point =>
        (horizontal ? point.current.offsetWidth : point.current.offsetHeight))

      dispatch({
        type: 'updatePointData',
        payload: { pointOffsetStarts, pointOffsetSizes },
      })
      dispatch({
        type: 'setFillLength',
        payload: {
          fillLength: `${pointOffsetStarts[currentPointIdIndex]}px`,
          fillTransition: false,
        },
      })
    }, [recalculate])

    return (
      <section className={cx('progress-bar', horizontal ? 'horizontal' : 'vertical')} ref={ref}>
        <div className={cx('progress-bar__meter-wrapper')}>
          <div className={cx('progress-bar__meter')} style={progressBarMeterStyle} />
          <div
            className={cx('progress-bar__meter', 'progress-bar__meter--fill', {
              'progress-bar__meter--transitioning': fillTransition,
            })}
            onTransitionEnd={() => {
              dispatch({ type: 'setFillIndex', payload: fillToIndex })
            }}
            style={barFillStyle}
          />
          <div className={cx('progress-bar__points')}>{points}</div>
        </div>
      </section>
    )
  },
)

ProgressBar.propTypes = {
  progressPoints: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    }),
  ).isRequired,
  currentPointId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  recalculate: PropTypes.string,
}

ProgressBar.defaultProps = {
  currentPointId: null,
  recalculate: null,
}

export default ProgressBar
