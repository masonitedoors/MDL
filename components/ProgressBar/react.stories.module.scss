@import "styles/index";

$min-width: 200px;

.progress-container {
  margin-left: 100px;

  input {
    margin: 15px 20px 15px 5px;
  }
}

.progress-point-container {
  margin: 0;
  padding: 0;
  list-style: none;
}

.progress-point {
  display: flex;
  align-items: flex-start;
  appearance: none;
  background-color: transparent;
  text-align: left;
  margin-left: $space-m;
  padding: 0;
  font-size: 12px;
  text-transform: uppercase;
  border: 0;
  color: $color-deepgray;
  font-weight: 700;
  min-width: $min-width;

  .horizontal & {
    position: relative;
    bottom: -30px;
    left: -50px;
    width: 120px;
    text-align: center;
  }
}

.change-size {
  background: none;
  border: none;
  margin: 0;
  padding: 0;
  cursor: pointer;
  text-decoration: underline;
  outline: none;
}

.big {
  height: 100px;
}
