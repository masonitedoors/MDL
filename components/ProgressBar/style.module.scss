@import "styles/index";

.progress-bar {
  position: relative;
  border-radius: 2px;

  &__meter-wrapper {
    position: absolute;

    .vertical & {
      left: 0;
      width: $space-s;
    }

    .horizontal & {
      top: 0;
      height: $space-s;
    }
  }

  &__meter {
    $distance-from-ends: $space-xs;
    position: absolute;
    margin: auto;
    background-color: $color-silver;

    .vertical & {
      width: $space-xs;
      left: 0;
      right: 0;
      top: $distance-from-ends;
      bottom: $distance-from-ends;
    }

    .horizontal & {
      height: $space-xs;
      left: $distance-from-ends;
      right: $distance-from-ends;
      top: 0;
      bottom: 0;
    }

    &--transitioning {
      transition: width 0.6s ease, height 0.6s ease;
    }

    &--fill {
      background-color: $color-limerick;

      .vertical & {
        bottom: initial;
      }

      .horizontal & {
        right: initial;
      }
    }
  }

  &__points {
    display: flex;

    .vertical & {
      flex-direction: column;
      height: 100%;
    }

    .horizontal & {
      flex-direction: row;
      width: 100%;
    }
  }

  &__point {
    position: relative;

    .vertical & {
      min-height: $space-xl;
      margin-bottom: $space-xs;
    }

    .horizontal & {
      min-width: $space-xl;
      margin-right: $space-xs;
    }

    &::after,
    &::before {
      @include circle($space-s);
      content: "";
      position: absolute;
      top: 0;
      left: 0;
      display: block;
      background-color: $color-silver;
      z-index: 1;
      transition: 200ms background-color ease-in;
    }

    &--active::before {
      background-color: rgba($color-night, 0.15);
      animation: grow-and-fade 400ms 1 normal;
      animation-timing-function: ease-in;
    }

    &--active::after {
      z-index: 1;
      background-color: $color-limerick;
      animation: pulse 400ms 1 normal;
      animation-timing-function: ease-in;
    }
  }

  &__point-label {
    left: $space-m;
  }
}

@keyframes grow-and-fade {
  to {
    $scale: 1;
    transform: scale3d($scale, $scale, $scale);
    opacity: 1;
  }

  from {
    $scale: 2.25;
    transform: scale3d($scale, $scale, $scale);
    opacity: 0;
  }
}

@keyframes pulse {
  0% {
    $scale: 1;
    transform: scale3d($scale, $scale, $scale);
  }

  50% {
    $scale: 1.2;
    transform: scale3d($scale, $scale, $scale);
  }

  100% {
    $scale: 1;
    transform: scale3d($scale, $scale, $scale);
  }
}
