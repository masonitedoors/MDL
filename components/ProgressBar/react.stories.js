import React, { useRef, useState } from 'react'
import classNames from 'classnames/bind'
import { storiesOf } from '@storybook/react'
import { withKnobs, boolean } from '@storybook/addon-knobs'
import ProgressBar from './react'
import storyStyles from './react.stories.module.scss'
import Readme from './react.README.md'

const cx = classNames.bind(storyStyles)

const Container = ({ horizontal, dynamic }) => {
  const ref = useRef()
  const [numElements, setNumElements] = useState(5)
  const [elementsExpanded, setElementsExpanded] = useState({})
  const [currentPointId, setCurrentPointId] = useState('2')

  const data = [...Array(numElements).keys()].map(num => ({
    id: `${num + 1}`,
    name: `Element ${num + 1}`,
  }))

  const toggleSize = id => () =>
    setElementsExpanded({
      ...elementsExpanded,
      [id]: !elementsExpanded[id],
    })

  const progressPoints = data.map(({ id, name }) => ({
    id,
    render: () => (
      <div className={cx('progress-point-container')}>
        <div className={cx('progress-point', { big: elementsExpanded[id] })}>
          {name}
          {dynamic && (
            <>
              &nbsp;(
              <button className={cx('change-size')} onClick={toggleSize(id)}>
                go {elementsExpanded[id] ? 'small' : 'big'}
              </button>
              )
            </>
          )}
        </div>
      </div>
    ),
  }))

  return (
    <div className={cx('progress-container', { horizontal })}>
      <label>
        Number of points:
        <input
          type="number"
          value={numElements}
          onChange={({ target }) => setNumElements(+target.value)}
        />
      </label>
      <label>
        Current points:
        <input
          type="number"
          value={currentPointId}
          onChange={({ target }) => setCurrentPointId(target.value)}
        />
      </label>
      <ProgressBar
        {...{
          progressPoints, currentPointId, ref, horizontal,
        }}
        recalculate={dynamic && elementsExpanded}
      />
    </div>
  )
}

storiesOf('ProgressBar', module)
  .addParameters({ readme: { sidebar: Readme } })
  .addDecorator(withKnobs)
  .add('Vertical/Dynamic', () => <Container dynamic />)
  .add('Horizontal/Static', () => <Container horizontal />)
