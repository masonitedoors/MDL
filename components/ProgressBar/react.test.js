import React from 'react'
import { shallow } from 'enzyme'
import {
  render, fireEvent, wait, waitForElement, cleanup,
} from '@testing-library/react'
import ProgressBar from './react'

describe('ProgressBar', () => {
  afterEach(cleanup)
  describe('with renders', () => {
    it('should render without crashing', () => {
      expect(shallow(<ProgressBar />)).toMatchSnapshot()
    })
  })
})
