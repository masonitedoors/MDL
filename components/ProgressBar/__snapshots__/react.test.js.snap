// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar with renders should render without crashing 1`] = `
<section
  className="progress-bar vertical"
>
  <div
    className="progress-bar__meter-wrapper"
  >
    <div
      className="progress-bar__meter"
      style={
        Object {
          "bottom": "NaNpx",
        }
      }
    />
    <div
      className="progress-bar__meter progress-bar__meter--fill progress-bar__meter--transitioning"
      onTransitionEnd={[Function]}
      style={
        Object {
          "height": "8px",
        }
      }
    />
    <div
      className="progress-bar__points"
    />
  </div>
</section>
`;
